File: app/src/main/java/com/kaltura/kflow/presentation/assetList/AssetListFragment.java
package com.kaltura.kflow.presentation.assetList;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kaltura.client.types.Asset;
import com.kaltura.client.types.ProgramAsset;
import com.kaltura.kflow.R;
import com.kaltura.kflow.presentation.player.PlayerFragment;
import com.kaltura.kflow.presentation.main.MainActivity;
import com.kaltura.kflow.utils.Utils;
import com.kaltura.playkit.PKMediaEntry;
import com.kaltura.playkit.providers.api.phoenix.APIDefines;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

/**
 * Created by alex_lytvynenko on 30.11.2018.
 */
public class AssetListFragment extends Fragment implements AssetListAdapter.OnAssetClickListener {

    private static final String ARG_ASSETS = "extra_assets";
    private static final String ARG_SCROLL_TO_LIVE = "extra_scroll_to_live";

    public static AssetListFragment newInstance(ArrayList<Asset> assets) {
        AssetListFragment assetListFragment = new AssetListFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(ARG_ASSETS, assets);
        assetListFragment.setArguments(bundle);
        return assetListFragment;
    }

    public static AssetListFragment newInstance(ArrayList<Asset> assets, boolean scrollToLive) {
        AssetListFragment assetListFragment = new AssetListFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(ARG_ASSETS, assets);
        bundle.putBoolean(ARG_SCROLL_TO_LIVE, scrollToLive);
        assetListFragment.setArguments(bundle);
        return assetListFragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_vod_list, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ((MainActivity) requireActivity()).getSupportActionBar().setTitle("Asset list");

        initList();
    }

    private void initList() {
        RecyclerView list = getView().findViewById(R.id.list);
        list.setHasFixedSize(true);
        list.setLayoutManager(new LinearLayoutManager(requireContext()));
        list.addItemDecoration(new DividerItemDecoration(requireContext(), LinearLayoutManager.VERTICAL));

        Bundle savedState = getArguments();
        ArrayList<Asset> assets = savedState != null ? (ArrayList<Asset>) savedState.getSerializable(ARG_ASSETS) : null;
        boolean scrollToLive = savedState != null && savedState.getBoolean(ARG_SCROLL_TO_LIVE);

        AssetListAdapter adapter = new AssetListAdapter(assets, this);
        list.setAdapter(adapter);

        if (scrollToLive && assets != null) {
            int liveAssetPosition = 0;
            for (Asset asset : assets) {
                if (asset instanceof ProgramAsset && Utils.isProgramInLive(asset)) {
                    liveAssetPosition = assets.indexOf(asset);
                    break;
                }
            }

            if (liveAssetPosition > 2)
                liveAssetPosition -= 3; // minus 3 items from the top, to move live asset to the middle of the screen

            list.scrollToPosition(liveAssetPosition);
        }
    }

    @Override
    public void onVodAssetClicked(Asset asset) {
        PlayerFragment playerFragment = PlayerFragment.newInstance(asset, false);
        requireActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.container, playerFragment)
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void onProgramAssetClicked(Asset asset, APIDefines.PlaybackContextType contextType) {
        if (!Utils.isProgramInFuture(asset)) {
            PlayerFragment playerFragment = PlayerFragment.newInstance(asset, false, contextType);
            requireActivity().getSupportFragmentManager().beginTransaction()
                    .replace(R.id.container, playerFragment)
                    .addToBackStack(null)
                    .commit();
        }
    }
}

File: app/src/main/java/com/kaltura/kflow/presentation/assetList/AssetListFragment.kt
package com.kaltura.kflow.presentation.assetList

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.LinearLayoutManager
import com.kaltura.client.types.Asset
import com.kaltura.client.types.ProgramAsset
import com.kaltura.kflow.R
import com.kaltura.kflow.databinding.FragmentAnonymousLoginBinding
import com.kaltura.kflow.databinding.FragmentVodListBinding
import com.kaltura.kflow.presentation.base.BaseFragment
import com.kaltura.kflow.presentation.extension.*
import org.koin.androidx.viewmodel.ext.android.viewModel

/**
 * Created by alex_lytvynenko on 30.11.2018.
 */
class AssetListFragment : BaseFragment(R.layout.fragment_vod_list) {

    private val viewModel: AssetListViewModel by viewModel()
    private val args: AssetListFragmentArgs by navArgs()
    private val adapter by lazy {
        AssetListAdapter(args.isShowActions).apply {
            vodClickListener = { asset ->
                val startPosition = args.watchedAssets?.firstOrNull { it.asset.id == asset.id }?.position
                        ?: 0
                navigate(AssetListFragmentDirections.navigateToPlayer(asset = asset, startPosition = startPosition))
            }
            programClickListener = { asset, contextType ->
                navigate(AssetListFragmentDirections.navigateToPlayer(asset = asset, playbackContextType = contextType.value))
            }
            reminderClickListener = { asset ->

                setAddReminderRequest(asset)
            }
        }
    }
    private var _binding: FragmentVodListBinding? = null
    private val binding get() = _binding!!
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentVodListBinding.inflate(inflater, container, false)
        val view = binding.root
        return view
    }
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.toolbar.setNavigationOnClickListener { activity?.onBackPressed() }
        initList()
    }

    private fun initList() {
        binding.list.setHasFixedSize(true)
        binding.list.layoutManager = LinearLayoutManager(requireContext())
        binding.list.layoutAnimation =
                if (adapter.assets.isEmpty()) AnimationUtils.loadLayoutAnimation(context, R.anim.item_layout_animation)
                else null

        binding.list.adapter = adapter
        if (args.watchedAssets == null) adapter.assets = args.assets!!
        else adapter.watchedAssets = args.watchedAssets!!

        if (args.isScrollToLive && args.assets!!.isNotEmpty()) {
            var liveAssetPosition = args.assets!!.indexOfFirst { it is ProgramAsset && it.isProgramInLive() }
            if (liveAssetPosition < 0) liveAssetPosition = 0
            if (liveAssetPosition > 2) liveAssetPosition -= 3 // minus 3 items from the top, to move live asset to the middle of the screen
            binding.list.scrollToPosition(liveAssetPosition)
        }
    }

    private fun setAddReminderRequest(program: Asset) {
        withInternetConnection {
            viewModel.makeAddReminderRequest(program)
        }
    }

    override fun subscribeUI() {
        observeResource(viewModel.reminderAddingEvent,
            error = {
                it.printStackTrace()
                longToast("Adding Reminder Error : $it")
            },
            success = {
                longToast("The following Reminder was successfully added : "+it)
            })
    }
}